### DISCUSSION TITLE
Ability to add structured data to head.tsx in /app

### DISCUSSION THREAD

**Discussion by trevorpfiz** [vercel/next.js#44629] [Ideas]
### Describe the feature you'd like to request

I would like to be able to add structured data to a Next.js 13 website using the appDir. [This was a related discussion in app directory feedback](https://github.com/vercel/next.js/discussions/41745#discussioncomment-4416236).

### Describe the solution you'd like

I would like to be able to add structured data to `head.tsx` in a similar way to this:

```
<script
      key="structured-data"
      type="application/ld+json"
      dangerouslySetInnerHTML={{ __html: JSON.stringify(structuredData) }}
    />
```

### Describe alternatives you've considered

```
<Script
    id='app-ld-json'
    type='application/ld+json'
    dangerouslySetInnerHTML={{
	    __html: JSON.stringify(content, null, '\t')
    }}
/>
```

It "works", but it is in the body and not the head.

**Comment by nitinTJ** [comment]
I have tried the same but it is working on my localhost but not on the production url. Although I can see it when I am inspecting the page but unable to find it in the page source. Any help will be appreciated.

`import React from "react";

interface ClientScriptProps {
  data: {};
}

function ClientScript({ data }: ClientScriptProps) {
  return (
    <section>
      {/* Add JSON-LD to your page */}
      <script
        type="application/ld+json"
        dangerouslySetInnerHTML={{ __html: JSON.stringify(data) }}
      />
      {/* ... */}
    </section>
  );
}

export default ClientScript;
`
I am using this component wherever I want to add JSON LD.

**Comment by roeean** [comment]
I have the same problem:

- Tried using /head.tsx, doesn't seem to work for each route.

- Tried using the <Head> tag, which doesn't seem to be supported in the new /app directory.

- Tried using next/script ([from this solution](https://stackoverflow.com/a/71747719/2419584)) but that seems to load the script on the client, I want it to be added at build-time or on the server.

- Tried using the new "[config-based metadata](https://beta.nextjs.org/docs/api-reference/metadata#config-based-metadata)", but it says that "script" tags are not supported.

**Reply by zain-sattar** [reply]
did you find any solution?

**Comment by todor0v** [comment]
If anyone is struggling with this, I found out that as per the [Google Search Central documentation](https://developers.google.com/search/docs/appearance/structured-data/intro-structured-data#supported-formats) and the [Rich Results Test](https://search.google.com/test/rich-results) it should be absolutely fine to have structured data in the body. This is also the [solution presented](https://nextjs.org/docs/app/building-your-application/optimizing/metadata#json-ld) by the Next.js documentation. 

If anyone needs dynamic data and still prefers the `<head>` approach, as do I, a possible solution is to make the request to that data twise. Once in the `layout.tsx` for the metadata and once in `page.tsx` for the remaining part. When you fetch the same data in a route more than once, React will [automatically dedupe the requests](https://nextjs.org/docs/app/building-your-application/caching#request-memoization) without affecting the performance.
